os/board/rtl8730e/src/component/usb/device_new/core/usbd_pcd.c

```diff
@@ -0,0 +1,2029 @@
+#if USBD_SWITCH_TO_ACTIVE_BY_SOF_INTR
+			/*
+				if the device wakes up the host remotely,
+				WKUINT intr will not be triggered
+				depend on sof intr to run the wakeup callback
+			*/
+			usbd_pcd_handle_wakeup_interrupt(pcd);
+#endif
+			usbd_pcd_handle_sof_interrupt(pcd);
+			USB_PCD_CLEAR_FLAG(USB_OTG_GINTSTS_SOF);
+		}
+
+		/* Handle EOPF Interrupt */
+		if (gintsts & (USB_OTG_GINTSTS_EOPF)) {
+			//RTK_LOGD(TAG, "EOPF\n");
+			usbd_pcd_handle_eopf_interrupt(pcd);
+			USB_PCD_CLEAR_FLAG(USB_OTG_GINTSTS_EOPF);
+		}
+
+		/* Handle Connection event Interrupt */
+		if (gintsts & (USB_OTG_GINTSTS_SRQINT)) {
+			RTK_LOGD(TAG, "SRQINT\n");
+			usbd_pcd_handle_srq_interrupt(pcd);
+			USB_PCD_CLEAR_FLAG(USB_OTG_GINTSTS_SRQINT);
+		}
+
+		/* Handle Disconnection event Interrupt */
+		if (gintsts & (USB_OTG_GINTSTS_OTGINT)) {
+			RTK_LOGD(TAG, "OTGINT\n");
+			usbd_pcd_handle_otg_interrupt(pcd);
+		}
+
+		usb_os_spinunlock(pcd->lock);
+	}
+}
+
+
+/* Exported functions --------------------------------------------------------*/
+
+/**
+  * @brief  Initialize PCD
+  * @param  dev: USB device instance
+  * @param  config: USB device configuration
+  * @retval HAL status
+  */
+USB_TEXT_SECTION
+u8 usbd_pcd_init(usb_dev_t *dev, usbd_config_t *config)
+{
+	u8 i;
+	u8 ret = HAL_OK;
+	usbd_pcd_ep_t *ep;
+	usbd_pcd_t *pcd = &usbd_pcd;
+
+	dev->pcd = &usbd_pcd;
+	pcd->dev = dev;
+
+	RTK_LOGI(TAG, "USB configurations:\n");
+	RTK_LOGI(TAG, "* speed: %d\n", config->speed);
+	RTK_LOGI(TAG, "* dma_enable: %d\n", config->dma_enable);
+	RTK_LOGI(TAG, "* isr_priority: %d\n", config->isr_priority);
+	RTK_LOGI(TAG, "* intr_use_ptx_fifo: %d\n", config->intr_use_ptx_fifo);
+	RTK_LOGI(TAG, "* ptx_fifo_first: %d\n", config->ptx_fifo_first);
+	RTK_LOGI(TAG, "* ext_intr_en: 0x%08X\n", config->ext_intr_en);
+	RTK_LOGI(TAG, "* nptx_max_epmis_cnt: %d\n", config->nptx_max_epmis_cnt);
+	for (i = 0U; i < USB_MAX_ENDPOINTS; i++) {
+		RTK_LOGI(TAG, "* nptx_max_err_cnt[%d]: %d\n", i, config->nptx_max_err_cnt[i]);
+	}
+
+	rtw_memcpy((void *)&pcd->config, (void *)config, sizeof(usbd_config_t));
+
+	for (i = 0U; i < USB_MAX_ENDPOINTS; i++) {
+		if (pcd->config.nptx_max_err_cnt[i] == 0U) {
+			pcd->config.nptx_max_err_cnt[i] = USBD_NPTX_DEF_ERR_CNT;
+		}
+	}
+
+	if (pcd->pcd_state == HAL_PCD_STATE_RESET) {
+		/* Allocate lock resource and initialize it */
+		pcd->lock = usb_os_spinlock_alloc();
+		if (pcd->lock == NULL) {
+			RTK_LOGE(TAG, "PCD lock init fail\n");
+			pcd->pcd_state = HAL_PCD_STATE_ERROR;
+			return HAL_ERR_MEM;
+		}
+
+		/* Init the low level hardware: GPIO, CLOCK, PHY... */
+		ret = usb_chip_init();
+		if (ret != HAL_OK) {
+			RTK_LOGE(TAG, "USB low-level init fail\n");
+			pcd->pcd_state = HAL_PCD_STATE_ERROR;
+			return ret;
+		}
+	}
+
+	pcd->pcd_state = HAL_PCD_STATE_BUSY;
+	pcd->phy_initialized = 0;
+	pcd->isr_initialized = 0;
+	pcd->setup = (u8 *)usb_os_mem_alloc(USBD_SETUP_PACKET_BUF_LEN);
+	if (pcd->setup == NULL) {
+		return HAL_ERR_MEM;
```

Comment:
Don't you need to free allocated memory before? I mean `pcd->lock`.
same as other returns

Reply:
Free allocated memory when error occurred